### PR TITLE
Bigint group by hash optimizations

### PR DESCRIPTION
**Pull request by skrzypo987**
## Description
A set of small optimizations for the BigIntGroupByHash class

> Is this change a fix, improvement, new feature, refactoring, or other?

improvement, refactoring
> Is this a change to the core query engine, a connector, client library, or the SPI interfaces? (be specific)

core query engine
> How would you describe this change to a non-technical end user or system administrator?

Nothing changes to a non-technical user
## Documentation

(x) No documentation is needed.
( ) Sufficient documentation is included in this PR.
( ) Documentation PR is available with #prnumber.
( ) Documentation issue #issuenumber is filed, and can be handled later.

## Release notes

( ) No release notes entries required.
( ) Release notes entries required with the following suggested text:

```markdown
# Section
* Improve performance of bigint aggregation
```

### CODE REVIEW

**Comment by radek-starburst** [comment]
Could you attach some macro-benchmark results?

File: core/trino-main/src/main/java/io/trino/operator/BigintGroupByHash.java [hunk 1]
```diff
@@ -302,22 +309,23 @@ private boolean tryRehash()
         int[] newGroupIds = new int[newCapacity];
         Arrays.fill(newGroupIds, -1);
 
-        for (int groupId = 0; groupId < nextGroupId; groupId++) {
-            if (groupId == nullGroupId) {
-                continue;
-            }
-            long value = valuesByGroupId.get(groupId);
+        for (int i = 0; i < values.length; i++) {
```
Comment (radek-starburst) [inline core/trino-main/src/main/java/io/trino/operator/BigintGroupByHash.java:312]: Why do we do it? We iterate by values instead of "groups". Then,we have one more sequential random access in the 314 line?

File: core/trino-main/src/main/java/io/trino/operator/BigintGroupByHash.java [hunk 2]
```diff
@@ -386,22 +392,30 @@ public AddPageWork(Block block)
         public boolean process()
         {
             int positionCount = block.getPositionCount();
-            checkState(lastPosition < positionCount, "position count out of bound");
 
-            // needRehash() == false indicates we have reached capacity boundary and a rehash is needed.
             // We can only proceed if tryRehash() successfully did a rehash.
-            if (needRehash() && !tryRehash()) {
-                return false;
+            while (nextGroupId + positionCount >= maxFill) {
```
Comment (radek-starburst) [inline core/trino-main/src/main/java/io/trino/operator/BigintGroupByHash.java:397]: Does that change make possible that some query fails now whileas that query does not fail before (when spill to disk is diabled)? 
I mean the situation when we have to process the last page and that page conatins 100k rows that are going to be assigned to one group.

File: core/trino-main/src/main/java/io/trino/operator/BigintGroupByHash.java [hunk 3]
```diff
@@ -302,22 +302,23 @@ private boolean tryRehash()
...
-            // find an empty slot for the address
-            int hashPosition = getHashPosition(value, newMask);
-            while (newGroupIds[hashPosition] != -1) {
-                hashPosition = (hashPosition + 1) & newMask;
-                hashCollisions++;
+        for (int i = 0; i < values.length; i++) {
+            long value = values[i];
+            int groupId = groupIds[i];
+
+            if (groupId != nullGroupId && groupId != -1) {
```
Comment (sopel39) [inline core/trino-main/src/main/java/io/trino/operator/BigintGroupByHash.java:316]: I don't think we ever assign `nullGroupId` to any `groupIds[i]`
```
    private int putIfAbsent(int position, Block block)
    {
        if (block.isNull(position)) {
            if (nullGroupId < 0) {
                // set null group id
                nullGroupId = nextGroupId++;
            }

            return nullGroupId;
        }
```

File: core/trino-main/src/main/java/io/trino/operator/BigintGroupByHash.java [hunk 4]
```diff
@@ -302,22 +302,23 @@ private boolean tryRehash()
...
-            }
-            long value = valuesByGroupId.get(groupId);
-
-            // find an empty slot for the address
-            int hashPosition = getHashPosition(value, newMask);
-            while (newGroupIds[hashPosition] != -1) {
-                hashPosition = (hashPosition + 1) & newMask;
-                hashCollisions++;
+        for (int i = 0; i < values.length; i++) {
+            long value = values[i];
```
Comment (sopel39) [inline core/trino-main/src/main/java/io/trino/operator/BigintGroupByHash.java:313]: load value in `if` loop

File: core/trino-main/src/main/java/io/trino/operator/BigintGroupByHash.java [hunk 5]
```diff
@@ -327,7 +325,7 @@ private boolean tryRehash()
         values = newValues;
         groupIds = newGroupIds;
 
-        this.valuesByGroupId.ensureCapacity(maxFill);
+        this.valuesByGroupId = Arrays.copyOf(valuesByGroupId, maxFill);
```
Comment (sopel39) [inline core/trino-main/src/main/java/io/trino/operator/BigintGroupByHash.java:337]: This is more expensive than `ensureCapacity` as `ensureCapacity` only adds segments. Can it be slower sometimes in practice? Multi channel still has `io.trino.operator.MultiChannelGroupByHash#groupAddressByGroupId`

**Review by sopel39 — COMMENTED** [review]
lgtm % comments up to `Replace block builder with a simple long[] `

**Comment by skrzypo987** [comment]
Comments addressed

File: core/trino-main/src/main/java/io/trino/operator/BigintGroupByHash.java (hunk 2 above)
Comment (radek-starburst) [reply core/trino-main/src/main/java/io/trino/operator/BigintGroupByHash.java:397]: It was just an example. Obviously we do not need a such specific case to reveal that issue. Actually, the last page can contain much less rows - like 2k or 300 or less. And it is not required for the last page to contain one group. The last page can contain a lot of group as well and query still could fail (whileas it succeed before).

File: core/trino-main/src/main/java/io/trino/operator/BigintGroupByHash.java (hunk 1 above)
Comment (radek-starburst) [reply core/trino-main/src/main/java/io/trino/operator/BigintGroupByHash.java:312]: I mean random. I cannot understand why we change it.

Comment (skrzypo987) [reply core/trino-main/src/main/java/io/trino/operator/BigintGroupByHash.java:312]: Previous approach was iterating over a group and then jumping around both `groupIds` and `values` arrays which was completely random.
Now we iterate over those two arrays in sequential order. Also `newGroupIds` whill be traversed quasi-sequentailly since it follow the same order as `groupIds` - it is ordered by hash.
Reach out to me offline if you need a more robust description.

File: core/trino-main/src/main/java/io/trino/operator/BigintGroupByHash.java (hunk 2 above)
Comment (skrzypo987) [reply core/trino-main/src/main/java/io/trino/operator/BigintGroupByHash.java:397]: I think that the requirements for that case are less probable than you think.
First of all in order to run out of memory query needs to be big.
So for a big enough query we have a case that:
- There is a lot of groups (so that the hash table is big enough). Let's assume 8M buckets (tpch/q18)
- The last page contains one or more groups that has not been seen before. Given that there already has been billions of incoming rows processed this is highly unlikely.
- The few added groups hit exactly the current limit. Given that the hash table is 4M buckets, there is a sort of 1/4M chance of that per any new group.
- Increasing the hash table size takes additional 4M*12B = 48MB. So now this number needs to hit exactly the memory limit. If the query is big enough to spawn millions of groups, the node/cluster probably has hundreds of GB of memory.

So, as you can see, the probability of hitting such an issue is minimal.
If you have a more likely example, please share it here.

File: core/trino-main/src/main/java/io/trino/operator/BigintGroupByHash.java (hunk 1 above)
Comment (radek-starburst) [reply core/trino-main/src/main/java/io/trino/operator/BigintGroupByHash.java:312]: Yeah, I get it.

File: core/trino-main/src/main/java/io/trino/operator/BigintGroupByHash.java (hunk 2 above)
Comment (radek-starburst) [reply core/trino-main/src/main/java/io/trino/operator/BigintGroupByHash.java:397]: I can imagine that probability of such case is minimal. Rather, the question for me is: is that acceptable for in comparison to the performance gain from that. Very possibly, yes - we can accept it. It is kind of tradeoff (for me).

File: core/trino-main/src/main/java/io/trino/operator/BigintGroupByHash.java [hunk 6]
```diff
@@ -52,6 +52,8 @@
     private static final float FILL_RATIO = 0.75f;
```
Comment (sopel39) [inline core/trino-main/src/main/java/io/trino/operator/BigintGroupByHash.java:52]: TBH I hardly see any difference in microbenchmarks beyond error margin. Not sure it's worth extra complication

File: core/trino-main/src/main/java/io/trino/operator/BigintGroupByHash.java [hunk 7]
```diff
@@ -376,8 +376,6 @@ private int registerGroupId(Block dictionary, int positionInDictionary)
     {
```
Comment (sopel39) [inline core/trino-main/src/main/java/io/trino/operator/BigintGroupByHash.java:383]: It seems it mostly helps for:
```
BenchmarkGroupByHashOnSimulatedData.groupBy             0   BIGINT_10K_GROUPS         ADD  avgt   10  10,671 ± 0,364  ns/op
BenchmarkGroupByHashOnSimulatedData.groupBy             0   BIGINT_10K_GROUPS  GET_GROUPS  avgt   10  13,108 ± 0,266  ns/op
```
why not other cases?
Does this makes a diff in practice

File: core/trino-main/src/main/java/io/trino/operator/BigintGroupByHash.java [hunk 8]
```diff
@@ -387,22 +385,19 @@ public AddPageWork(Block block)
         public boolean process()
         {
             int positionCount = block.getPositionCount();
-            checkState(lastPosition < positionCount, "position count out of bound");
```
[the remark this answers is not in the document]
Comment (skrzypo987) [reply core/trino-main/src/main/java/io/trino/operator/BigintGroupByHash.java:389]: If the page is full RLE then it will not land here. At least one block must be non-RLE. Is there a possibility of that kind of page to be that huge, e.g >10k positions? I assume that page is by-definition a chunk of data that can be processed atomically.

File: core/trino-main/src/main/java/io/trino/operator/BigintGroupByHash.java (hunk 7 above)
Comment (skrzypo987) [reply core/trino-main/src/main/java/io/trino/operator/BigintGroupByHash.java:376]: It's a slight change. I don't think we should expect huge gains here. It's rather a simplification than a performance trick.

File: core/trino-main/src/main/java/io/trino/operator/BigintGroupByHash.java (hunk 6 above)
Comment (skrzypo987) [reply core/trino-main/src/main/java/io/trino/operator/BigintGroupByHash.java]: I guess you are right. I'll drop the commit

File: core/trino-main/src/main/java/io/trino/operator/BigintGroupByHash.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (sopel39) [reply core/trino-main/src/main/java/io/trino/operator/BigintGroupByHash.java:389]: >  Is there a possibility of that kind of page to be that huge, e.g >10k positions?

You are betting that every place in code will adhere to these rules and not produce extremely large pages. Additionally, we don't have a control over all connectors, which can potentially produce large pages too. It's better to be defensive here

[the remark this answers is not in the document]
Comment (skrzypo987) [reply core/trino-main/src/main/java/io/trino/operator/BigintGroupByHash.java:389]: I added batching. No performance difference

File: core/trino-main/src/test/java/io/trino/operator/BenchmarkGroupByHashOnSimulatedData.java [hunk 9]
```diff
@@ -0,0 +1,626 @@
...
+    public enum AggregationDefinition
+    {
+        BIGINT_2_GROUPS(new ChannelDefinition(ColumnType.BIGINT, 2)),
+        BIGINT_10_GROUPS(new ChannelDefinition(ColumnType.BIGINT, 10)),
+        BIGINT_1K_GROUPS(new ChannelDefinition(ColumnType.BIGINT, 1000)),
+        BIGINT_10K_GROUPS(new ChannelDefinition(ColumnType.BIGINT, 10_000)),
+        BIGINT_100K_GROUPS(new ChannelDefinition(ColumnType.BIGINT, 100_000)),
+        BIGINT_1M_GROUPS(new ChannelDefinition(ColumnType.BIGINT, 1_000_000)),
+        BIGINT_10M_GROUPS(new ChannelDefinition(ColumnType.BIGINT, 10_000_000)),
+        BIGINT_2_GROUPS_1_SMALL_DICTIONARY(new ChannelDefinition(ColumnType.BIGINT, 2, 1, 50)),
```
Comment (sopel39) [inline core/trino-main/src/test/java/io/trino/operator/BenchmarkGroupByHashOnSimulatedData.java:331]: nit: add textual description to these enums (can be separate PR). It's hard to understand what they benchmark after some time

**Review by sopel39 — APPROVED** [review]

File: core/trino-main/src/main/java/io/trino/operator/BigintGroupByHash.java [hunk 10]
```diff
@@ -34,20 +34,23 @@
...
 import static it.unimi.dsi.fastutil.HashCommon.murmurHash3;
+import static java.lang.Math.min;
 import static java.lang.Math.toIntExact;
 import static java.util.Objects.requireNonNull;
 
 public class BigintGroupByHash
         implements GroupByHash
 {
     private static final int INSTANCE_SIZE = ClassLayout.parseClass(BigintGroupByHash.class).instanceSize();
+    private static final int BATCH_SIZE = 1024;
```
Comment (sopel39) [inline core/trino-main/src/main/java/io/trino/operator/BigintGroupByHash.java:53]: nit: this is bigger than expected groups for `io.trino.sql.planner.LocalExecutionPlanner.Visitor#visitTableWriter` and `io.trino.sql.planner.LocalExecutionPlanner.Visitor#visitTableFinish` (200)

**Review by sopel39 — APPROVED** [review]

File: core/trino-main/src/main/java/io/trino/operator/BigintGroupByHash.java (hunk 10 above)
Comment (skrzypo987) [reply core/trino-main/src/main/java/io/trino/operator/BigintGroupByHash.java:53]: Expected number of groups is hardcoded as `10000` so I don't think that matters.

File: core/trino-main/src/test/java/io/trino/operator/BenchmarkGroupByHashOnSimulatedData.java (hunk 9 above)
Comment (skrzypo987) [reply core/trino-main/src/test/java/io/trino/operator/BenchmarkGroupByHashOnSimulatedData.java:331]: https://github.com/trinodb/trino/pull/12170
